### PR TITLE
Bug 1766984: trustedCA bundle support for alertmanager oauth proxy

### PR DESCRIPTION
**Pull request by paulfantom**
We need to mount trustedCA bundle into a place that is understandable by oauth-proxy and from where a bundle can be picked up. This is similar to how we did it for telemeter.

Fix is based on @gfvirga [comment](https://github.com/openshift/cluster-monitoring-operator/pull/448#issuecomment-547051573) and further investigation with @sur

@LiliC this might also solve alertmanager problems with CA bundle.

Part 1/3 of fixing oauth-proxy problems with trustedCA budle :)

### CODE REVIEW

**Comment by s-urbaniak** [comment]
@brancz do you mind to have a look quickly? Is `/etc/pki/ca-trust/extracted/pem/` something we cannot use due to alertmanager internals?

**Comment by brancz** [comment]
I feel like we tried exactly this and the Alertmanager internal gossip communication didn't work anymore because of it. If we can double check that this is not the case, then I'm happy with going ahead with this.

**Comment by paulfantom** [comment]
> If we can double check that this is not the case

Launching cluster now :)

---

> Alertmanager internal gossip communication didn't work anymore because of it

Should that result in degraded CMO?

**Comment by gfvirga** [comment]
@paulfantom , I saw you had a PR for grafana, are you planning one for Prometheus and logging as well?


I can confirm that the telemeter is working as expected, curl is not complaining about the self signed certificate it's using the bundle.

```
oc rsh telemeter-client-587877fd9c-vwvkv
Defaulting container name to telemeter-client.
Use 'oc describe pod/telemeter-client-587877fd9c-vwvkv -n openshift-monitoring' to see all of the containers in this pod.
sh-4.2$ curl https://oauth-openshift.apps.ose.example.com/
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "forbidden: User \"system:anonymous\" cannot get path \"/\"",
  "reason": "Forbidden",
  "details": {
    
  },
  "code": 403
}
```

Telemeter is also mouting the bundle under pem.
```
      volumeMounts:
        - name: telemeter-trusted-ca-bundle
          readOnly: true
          mountPath: /etc/pki/ca-trust/extracted/pem/
```

**Comment by s-urbaniak** [comment]
@gfvirga yes, we plan to add a follow-up PR for prometheus. What do you mean with logging? (it is not part of the monitoring stack). Did you recognize issues there as well?

**Comment by paulfantom** [comment]
Putting on hold as we need to test it more thoroughly.
/hold

---

@gfvirga yes, I plan on doing a fix for prometheus too, but this will first need an upstream change in prometheus-operator. As for logging, this is not in monitoring-team hands, best would be to create a second BZ and assign it to logging component.

**Comment by paulfantom** [comment]
I tested it manually with a proxy and alertmanager didn't have any problems with gossip network, CA was mounted correctly and it all seemed to work. Removing hold and waiting for reviews

/hold cancel
/cc @LiliC @s-urbaniak

File: pkg/manifests/manifests.go [hunk 1]
```diff
@@ -364,8 +364,10 @@ func (f *Factory) AlertmanagerMain(host string, trustedCABundleCM *v1.ConfigMap)
 
 	if trustedCABundleCM != nil {
 		volumeName := "alertmanager-trusted-ca-bundle"
-		a.Spec.VolumeMounts = append(a.Spec.VolumeMounts, trustedCABundleVolumeMount(volumeName, "/etc/pki/alertmanager-ca-bundle/"))
+		volumePath := "/etc/pki/ca-trust/extracted/pem/"
+		a.Spec.VolumeMounts = append(a.Spec.VolumeMounts, trustedCABundleVolumeMount(volumeName, volumePath))
 		a.Spec.Volumes = append(a.Spec.Volumes, trustedCABundleVolume(trustedCABundleCM.Name, volumeName))
+		a.Spec.Containers[0].VolumeMounts = append(a.Spec.Containers[0].VolumeMounts, trustedCABundleVolumeMount(volumeName, volumePath))
```
Comment (LiliC) [inline pkg/manifests/manifests.go]: We add this to just the first container?

**Comment by gfvirga** [comment]
> @gfvirga yes, we plan to add a follow-up PR for prometheus. What do you mean with logging? (it is not part of the monitoring stack). Did you recognize issues there as well?

@s-urbaniak , I opened an issue under logging too. The same thing happens there. https://github.com/openshift/cluster-logging-operator/issues/261#issuecomment-547700734

I feel like I will have to just get verisign SSL certs for my cluster to fully work. :(

File: pkg/manifests/manifests.go (hunk 1 above)
Comment (paulfantom) [reply pkg/manifests/manifests.go]: updated to go through all containers

**Comment by paulfantom** [comment]
rebased to include grafana oauth-proxy for easier manual testing

File: pkg/manifests/manifests.go (hunk 1 above)
Comment (paulfantom) [reply pkg/manifests/manifests.go]: I was looking more into it. We are adding it into spec of alertmanager CR. This one has only 1 additional container - oauth-proxy. So it shouldn't change anything.

**Comment by paulfantom** [comment]
After a lot of investigation, it appears we are blocked by apiserver bug regarding validation of CRDs (https://github.com/kubernetes/kubernetes/issues/84880). On a plus side, @sttts is involved in resolving it and there is already a potential fix in https://github.com/go-openapi/validate/pull/119

**Comment by LiliC** [comment]
@paulfantom I think I hit this bug as well! For some reason we thought it was because of the way we do generation in prom-operaotr 🤔 @s-urbaniak

**Comment by s-urbaniak** [comment]
@LiliC for this specific case it is indeed an upstream bug in openapi validation. I rabbit-holed this together with @sttts. The fix is upstream in https://github.com/kubernetes/kubernetes/pull/85223 and will most likely be bumped in origin, then we can proceed here.

**Comment by LiliC** [comment]
/test e2e-aws
This was the test failure, seems related-ish:
 
> fail [github.com/openshift/origin/test/extended/prometheus/prometheus_builds.go:134]: Expected
    <map[string]error | len:1>: {
        "prometheus_rule_evaluation_failures_total >= 1": {
            s: "promQL query: prometheus_rule_evaluation_failures_total >= 1 had reported incorrect results: prometheus_rule_evaluation_failures_total{endpoint=\"web\", instance=\"10.129.2.9:9091\", job=\"prometheus-k8s\", namespace=\"openshift-monitoring\", pod=\"prometheus-k8s-0\", service=\"prometheus-k8s\"} => 8 @[1574116495.841]",
        },
    }
to be empty

**Comment by s-urbaniak** [comment]
looking good so far modulo the investigation if https://github.com/openshift/cluster-monitoring-operator/pull/529#issuecomment-555327382 is related.

**Comment by paulfantom** [comment]
Created a follow-up BZ regarding rule evaluation failures - https://bugzilla.redhat.com/show_bug.cgi?id=1774010

**Pull request merged**